lms/djangoapps/mobile_api/users/serializers.py

```diff
@@ -141,6 +162,101 @@ class Meta:
         lookup_field = 'username'
 
 
+class CourseEnrollmentSerializerModifiedForPrimary(CourseEnrollmentSerializer):
+    """
+    Serializes CourseEnrollment models for API v4.
+
+    Adds `course_status` field into serializer data.
+    """
+
+    course_status = serializers.SerializerMethodField()
+    course_progress = serializers.SerializerMethodField()
+    course_assignments = serializers.SerializerMethodField()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.course = modulestore().get_course(self.instance.course.id)
+
+    def get_course_status(self, model: CourseEnrollment) -> Optional[Dict[str, List[str]]]:
```

Comment:
Seems like most of this logic is already written in this [view](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/mobile_api/users/views.py#L90).
Can we reuse this logic without duplicating it?

Reply:
Yes, I see. This is exactly what we did initially. We tried to use methods from this view. We used functions from this method `_last_visited_block_path`. We also moved the repetitive logic into a separate function. But our QAs found out that the method `_get_course_info` returns the wrong path to the last visited block. So, we decided to create new method that returns correct path to the last visited block and don't use logic from `_last_visited_block_path` method.

Maybe some problems are in this function:
```
def get_current_child(xblock, min_depth=None, requested_child=None):
    """
    Get the xblock.position's display item of an xblock that has a position and
    children.  If xblock has no position or is out of bounds, return the first
    child with children of min_depth.
```
We may not understand how this function is supposed to work. But the path to last visited block that is built using this function is not correct

Reply:
@KyryloKireiev Does `get_current_child` return the wrong block, or is the right block with the wrong path?

Reply:
@kdmccormick `get_current_child` function is used in `UserCourseStatus._last_visited_block_path` method. This method `_last_visited_block_path` returns incorrect path to the last visited block. Maybe `get_current_child` function works correctly. But I think it is not correct to use it to find path to the last visited block.